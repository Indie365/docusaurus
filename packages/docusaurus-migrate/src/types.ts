/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export type RawData = {
  header?: string;
  content?: string;
};

export type Data = {
  metadata: {[key: string]: string};
  rawContent: string;
};
export interface DocusaurusConfig {
  baseUrl: string;
  favicon: string;
  tagline?: string;
  title: string;
  url: string;
  organizationName?: string;
  projectName?: string;
  githubHost?: string;
  plugins: Array<[string, {[key: string]: any}]>;
  themes?: [];
  presets: [
    [
      string,
      {
        docs: {[key: string]: any};
        blog: {[key: string]: any};
        theme: {[key: string]: any};
      },
    ],
  ];
  themeConfig: {
    navbar: {
      title?: string;
      logo: {
        src?: string;
      };
      image?: string;
      links: Array<object>;
    };
    footer: {
      links: Array<{
        title: string;
        items: Array<{
          label: string;
          to: string;
        }>;
      }>;
      copyright?: string;
      logo: {
        src?: string;
      };
    };
    algolia?: object;
  };
  customFields: {
    [key: string]: unknown;
  };
  scripts?: (
    | string
    | {
        src: string;
        [key: string]: unknown;
      }
  )[];
  stylesheets?: (
    | string
    | {
        href: string;
        [key: string]: unknown;
      }
  )[];
}

export type Config = {
  title?: string;
  tagline?: string;
  url?: string;
  baseUrl?: string;
  defaultVersionShown?: string;
  organizationName?: string;
  projectName?: string;
  noIndex?: string;
  headerLinks?: Array<any>;
  headerIcon?: string;
  favicon?: string;
  colors?: any;
  copyright?: string;
  editUrl?: string;
  users?: Array<object>;
  disableHeaderTitle?: string;
  disableTitleTagline?: string;
  separateCss?: Array<object>;
  footerIcon?: string;
  translationRecruitingLink?: string;
  algolia?: object;
  gaTrackingId?: string;
  highlight?: object;
  markdownPlugins?: Array<() => void>;
  scripts?: Array<{src: string; [key: string]: any} | string>;
  stylesheets?: Array<{href: string; [key: string]: any} | string>;
  facebookAppId?: string;
  facebookComments?: true;
  facebookPixelId?: string;
  twitter?: string;
  twitterUsername?: string;
  twitterImage?: string;
  ogImage?: string;
  cleanUrl?: boolean;
  scrollToTop?: boolean;
  scrollToTopOptions?: object;
};
