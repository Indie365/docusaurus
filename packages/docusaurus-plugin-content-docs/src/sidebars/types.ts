/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {
  DocMetadataBase,
  VersionMetadata,
  NumberPrefixParser,
  SidebarOptions,
} from '../types';

export type SidebarItemBase = {
  className?: string;
  customProps?: Record<string, unknown>;
};

export type SidebarItemDoc = SidebarItemBase & {
  type: 'doc' | 'ref';
  label?: string;
  id: string;
};

export type SidebarItemLink = SidebarItemBase & {
  type: 'link';
  href: string;
  label: string;
};

export type SidebarItemAutogenerated = SidebarItemBase & {
  type: 'autogenerated';
  dirName: string;
};

// The user-input configuration, before normalization
export type SidebarItemCategoryConfig = SidebarItemBase & {
  type: 'category';
  label: string;
  items: SidebarItemConfig[];
  collapsed?: boolean;
  collapsible?: boolean;
};

export type SidebarItemConfig =
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemAutogenerated
  | SidebarItemCategoryConfig
  | string
  | SidebarCategoriesShorthand;

export type SidebarCategoriesShorthand = {
  [sidebarCategory: string]: SidebarItemConfig[];
};

export type SidebarConfig = SidebarCategoriesShorthand | SidebarItemConfig[];

// Sidebar given by user that is not normalized yet. e.g: sidebars.json
export type SidebarsConfig = {
  [sidebarId: string]: SidebarConfig;
};

// Still has 'autogenerated', which is handled in processing
export type NormalizedSidebarItemCategory = SidebarItemBase & {
  type: 'category';
  label: string;
  items: NormalizedSidebarItem[];
  collapsed: boolean;
  collapsible: boolean;
};

export type NormalizedSidebarItem =
  | SidebarItemDoc
  | SidebarItemLink
  | NormalizedSidebarItemCategory
  | SidebarItemAutogenerated;

export type NormalizedSidebar = NormalizedSidebarItem[];
export type NormalizedSidebars = {
  [sidebarId: string]: NormalizedSidebar;
};

export type SidebarItemCategory = SidebarItemBase & {
  type: 'category';
  label: string;
  items: SidebarItem[];
  collapsed: boolean;
  collapsible: boolean;
};

export type SidebarItem =
  | SidebarItemDoc
  | SidebarItemLink
  | SidebarItemCategory;

export type Sidebar = SidebarItem[];
export type SidebarItemType = SidebarItem['type'];
export type Sidebars = {
  [sidebarId: string]: Sidebar;
};

// Reduce API surface for options.sidebarItemsGenerator
// The user-provided generator fn should receive only a subset of metadatas
// A change to any of these metadatas can be considered as a breaking change
export type SidebarItemsGeneratorDoc = Pick<
  DocMetadataBase,
  'id' | 'frontMatter' | 'source' | 'sourceDirName' | 'sidebarPosition'
>;
export type SidebarItemsGeneratorVersion = Pick<
  VersionMetadata,
  'versionName' | 'contentPath'
>;

export type SidebarItemsGeneratorArgs = {
  item: SidebarItemAutogenerated;
  version: SidebarItemsGeneratorVersion;
  docs: SidebarItemsGeneratorDoc[];
  numberPrefixParser: NumberPrefixParser;
  options: SidebarOptions;
};
export type SidebarItemsGenerator = (
  generatorArgs: SidebarItemsGeneratorArgs,
) => Promise<SidebarItem[]>;

// Also inject the default generator to conveniently wrap/enhance/sort the default sidebar gen logic
// see https://github.com/facebook/docusaurus/issues/4640#issuecomment-822292320
export type SidebarItemsGeneratorOptionArgs = {
  defaultSidebarItemsGenerator: SidebarItemsGenerator;
} & SidebarItemsGeneratorArgs;
export type SidebarItemsGeneratorOption = (
  generatorArgs: SidebarItemsGeneratorOptionArgs,
) => Promise<SidebarItem[]>;

export type OrderMetadata = {
  previous?: string;
  next?: string;
  sidebar?: string;
};
