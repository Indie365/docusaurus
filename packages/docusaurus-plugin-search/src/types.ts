/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

export interface Node {
  /**
   * The variant of a node.
   */
  type: string;

  /**
   * Information from the ecosystem.
   */
  data?: unknown;

  /**
   * Location of a node in a source document.
   * Must not be present if a node is generated.
   */
  position?: unknown;

  [key: string]: unknown;
}

export interface DataNode {
  type: string;
  anchor?: string;
  body: string;
  heading?: string;
}

export interface pluginOptions {
  include: Array<string>;
  exclude: Array<string>;
}
