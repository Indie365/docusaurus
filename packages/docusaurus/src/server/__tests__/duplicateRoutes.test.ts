/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import {
  getAllDuplicateRoutes,
  getDuplicateRoutesMessage,
} from '../duplicateRoutes';
import {RouteConfig} from '@docusaurus/types';

describe('duplicateRoutes', () => {
  test('getDuplicateRoutesMessage', () => {
    const message = getDuplicateRoutesMessage([
      '/',
      '/',
      '/blog',
      '/doc/search',
    ]);
    expect(message).toMatchSnapshot();
  });

  test('getAllDuplicateRoutes', () => {
    const routes: RouteConfig[] = [
      {
        path: '/',
        component: '',
        routes: [
          {path: '/search', component: ''},
          {path: '/sameDoc', component: ''},
        ],
      },
      {
        path: '/',
        component: '',
        routes: [
          {path: '/search', component: ''},
          {path: '/sameDoc', component: ''},
          {path: '/uniqueDoc', component: ''},
        ],
      },
    ];
    expect(getAllDuplicateRoutes(routes)).toMatchSnapshot();
  });
});
