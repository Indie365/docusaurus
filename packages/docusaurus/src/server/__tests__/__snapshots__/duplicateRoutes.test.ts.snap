// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duplicateRoutes getAllDuplicateRoutes 1`] = `
Array [
  "/",
  "/search",
  "/sameDoc",
]
`;

exports[`duplicateRoutes getDuplicateRoutesMessage 1`] = `
"Attempting to create page at /, but a page already exists at this route
Attempting to create page at /, but a page already exists at this route
Attempting to create page at /blog, but a page already exists at this route
Attempting to create page at /doc/search, but a page already exists at this route"
`;
